R/morphemepiece.R

```diff
@@ -196,38 +305,243 @@ magrittr::`%>%`
 #' Tokenize Sequence with Morpheme Pieces
 #'
 #' Given a single sequence of text and a morphemepiece vocabulary, tokenizes the
-#' text. (This is *not* fully functional yet. It's mainly a placeholder so that
-#' the package exports something.)
-#' 
+#' text.  
 #'
-#' @inheritParams .mp_tokenize_word
+#' @inheritParams .mp_tokenize_word_lookup
 #' @param text Character scalar; text to tokenize.
 #'
 #' @return A character vector of tokenized text (later, this should be a named
 #' integer vector, as in the wordpiece package.)
 #' @export
 morphemepiece_tokenize <- function(text,
-                                   vocab) {
-
+                                   vocab,
+                                   lookup,
+                                   unk_token = "[UNK]",
+                                   max_chars = 100) {
+  is_cased <- attr(vocab, "is_cased")
+  if (!is_cased) {
     text <- tolower(text)
+  }
+  
+  text <- .convert_to_unicode(text)
+  text <- .clean_text(text)
+  text <- .tokenize_chinese_chars(text)
+  text <- .strip_accents(text)
+  text <- .split_on_punc(text)
+  text <- purrr::map(wordpiece:::.whitespace_tokenize(text),
+                     .f = .mp_tokenize_word_lookup, 
+                     vocab = vocab, 
+                     lookup = lookup)
+  text <- unlist(text)
+  return(text)
+  #  For testing on datascience:
+  # mp_vocab <- load_or_retrieve_vocab("/shared/morphemepiece_vocabs/mp_vocab_big.txt")
+  # mp_lookup <- load_or_retrieve_lookup("/shared/morphemepiece_vocabs/mp_lookup_big.txt")
+  # morphemepiece_tokenize("I love tacos! And prexxxxationings", mp_vocab, mp_lookup)
+}
+
+
+# load_vocab --------------------------------------------------------------
+
+
+#' Load a vocabulary file
+#'
+#' @param vocab_file path to vocabulary file. File is assumed to be a text file,
+#'   with one token per line, with the line number (starting at zero)
+#'   corresponding to the index of that token in the vocabulary.
+#'
+#' @return The vocab as a named integer vector. Names are tokens in vocabulary,
+#'   values are integer indices. The casedness of the vocabulary is inferred
+#'   and attached as the "is_cased" attribute.
+#'
+#'   Note that from the perspective of a neural net, the numeric indices *are*
+#'   the tokens, and the mapping from token to index is fixed. If we changed the
+#'   indexing, it would break any pre-trained models using that vocabulary. This
+#'   is why the vocabulary is stored as a named integer vector, and why it
+#'   starts with index zero.
+#'
+#' @export
+#'
+#' @examples
+#' # todo, after I make tiny sample vocab to include
+load_vocab <- function(vocab_file) {
```

Comment:
Should we go ahead and add the wordpiece dependency now, instead of this function? I know we have other things we'll want from wordpiece later, but it feels easier to import this directly right away (to avoid fixes in 2 places even for a moment).

Reply:
eventually we will figure out the overlapping code between this and wordpiece, but this function isn't exactly the same as the one in wordpiece.